File: test/end-to-end/VectorClass/C#/Vector.cs
using System;
using System.Collections.Generic;

namespace VectorClass
{
    public class Vector<T>
    {
        private T[] data;
        private int capacity;
        private int length;

        public Vector()
        {
            this.capacity = 0;
            this.data = new T[0];
            this.length = 0;
        }

        public T At(int index)
        {
            if (index >= this.length)
            {
                throw new Exception(string.Format("Index out of bounds: {0} is greater than {1}.", index, this.length));
            }

            return this.data[index];
        }

        public int GetCapacity()
        {
            return this.capacity;
        }

        public int GetLength()
        {
            return this.length;
        }

        public T GetFirst()
        {
            if (this.capacity == 0)
            {
                throw new Exception("Cannot get first from empty vector.");
            }

            return this.data[0];
        }

        public T GetLast()
        {
            if (this.capacity == 0)
            {
                throw new Exception("Cannot get last from empty vector.");
            }

            return this.data[this.length - 1];
        }

        public void EnsureCapacity(int capacity)
        {
            if (capacity <= this.capacity)
            {
                return;
            }

            int newCapacity = (int)(Math.Ceiling((capacity / 2.0)) * 3);
            T[] oldData = this.data;
            this.capacity = newCapacity;
            this.data = new T[newCapacity];

            for (int i = 0; i < oldData.Length; i += 1)
            {
                this.data[i] = oldData[i];
            }
        }

        public void Push(T item)
        {
            this.EnsureCapacity(this.length + 1);

            this.data[this.length] = item;
            this.length += 1;
        }

        public void Resize(int length)
        {
            if (length <= this.length)
            {
                this.length = length;
                return;
            }

            this.EnsureCapacity(length);
            this.length = length;
        }

        public T[] ToArray()
        {
            T[] array = new T[this.length];

            for (int i = 0; i < this.length; i += 1)
            {
                array[i] = this.data[i];
            }

            return array;
        }

        public List<T> ToList()
        {
            List<T> list = new List<T>(this.length);

            for (int i = 0; i < this.length; i += 1)
            {
                list[i] = this.data[i];
            }

            return list;
        }

        public HashSet<T> ToSet()
        {
            HashSet<T> individuals = new HashSet<T>();

            for (int i = 0; i < this.length; i += 1)
            {
                individuals.Add(this.data[i]);
            }

            return individuals;
        }
    }
}

File: test/end-to-end/VectorClass/PHP/vector.php
require "set"

<?php
class Vector
    attr_accessor :data;
    attr_accessor :capacity;
    attr_accessor :length;

    function __constructor()
        self.capacity = 0;
        self.data = [];
        self.length = 0;
    end

    def at(index)
        if index >= self.length
            raise Exception.new('Index out of bounds: ' . index . ' is greater than ' . self.length . '.');
        end

        return self.data[index];
    end

    def get_capacity()
        return self.capacity;
    end

    def get_length()
        return self.length;
    end

    def get_first()
        if self.capacity == 0
            raise Exception.new("Cannot get first from empty vector.");
        end

        return self.data[0];
    end

    def get_last()
        if self.capacity == 0
            raise Exception.new("Cannot get last from empty vector.");
        end

        return self.data[self.length - 1];
    end

    def ensure_capacity(capacity)
        if capacity <= self.capacity
            return;
        end

        $newCapacity
        $oldData
        self.capacity = newCapacity;
        self.data = Array.new(newCapacity);

        for i in 0...count(oldData)
            self.data[i] = oldData[i];
        end
    end

    def push(item)
        self.ensure_capacity(self.length + 1);

        self.data[self.length] = item;
        self.length += 1;
    end

    def resize(length)
        if length <= self.length
            self.length = length;
            return;
        end

        self.ensure_capacity(length);
        self.length = length;
    end

    def to_array()
        $array

        for i in 0...self.length
            array[i] = self.data[i];
        end

        return array;
    end

    def to_list()
        $list

        for i in 0...self.length
            list[i] = self.data[i];
        end

        return list;
    end

    def to_set()
        $individuals

        for i in 0...self.length
            individuals.add(self.data[i]);
        end

        return individuals;
    end
end
